### DISCUSSION TITLE
How to implement manual toggle for selection modes and fixing/adding extra implementation for drag and drop and reordering?

### DISCUSSION THREAD

**Discussion by Rhitomaic** [yasirkula/UnityRuntimeInspector#61] [Q&A] [answered]
So I have a level editor that uses this library's hierarchy but I don't know how to properly set this up yet.
I want to have a, probably secondary enum that can set the default selection mode because the level editor is also made for Android and they cannot really hold shift or ctrl, I want to control that enum with a button and I already have a script to be able to control an enum with a button. I just want to know how to properly implement it.

For the second one is extra implementation for drag and drop,
Since all objects in the scene is put in a pseudo scene, I want to be able to restrict certain objects to be parented based on conditions, actually reorder the game objects in the scene by setting it's sibling index relative to whatever is above or below the hierarchy field, and remove items from pseudo scenes if it just so happens to be parented to a child.

Thank you for your help! :D

**Comment by yasirkula** [comment, marked as answer]
1) You can enable runtimeHierarchy.MultiSelectionToggleSelectionMode via your script.
2) Can you elaborate your setup? Are all those root objects child of the same Transform?

**Reply by Rhitomaic** [reply]
1. What if I want to have three types of selection mode like, area selection where clicking two spaced items will select both and in between, multiple selection where you can just click to select and deselect anything without deselecting previous selection, and just... normal? or single selection where you're only allowed to select a single object

2. I have two pseudo scenes, I was using the old version where it has no option to attach a host transform so I just add all the objects manually when an object is created. The two pseudo scenes are:
- Base: Objects that aren't meant to be moved around in the hierarchy or deleted, they're supposed to always exist and in the same parent
- Environment: Objects in this pseudo-scene can be modified, deleted, added more, parented to objects in environment and base

These pseudo-scenes are actually just a representation of two objects, and what I'm doing is manually adding the child to those pseudo-scenes. There's a catch though, I have an object that is not in the Base GameObject and I need to add that to the Base pseudo-scene somehow, so yeah, only one object that isn't parented yet it has to show anyway in the Base pseudo-scene

Also I have just updated the library, so I'll try making changes when waiting for a reply

**Reply by yasirkula** [reply]
1. Area selection isn't implemented on Android so you'll need to implement it yourself. The other two already exist (`MultiSelectionToggleSelectionMode`) and multi-selection can also be toggled via `AllowMultiSelection` You could also set `PointerLongPressAction` to `LongPressAction.ShowMultiSelectionTogglesThenCreateDraggedReferenceItem` to intuitively switch to multi-selection mode on Android.
2. Try adding PseudoSceneSourceTransform component to Base and Environment and let that component handle adding/removing objects to/from their corresponding pseudo-scenes. You can still add the other object manually via `AddToPseudoScene`.

**Reply by Rhitomaic** [reply]
1. I'll discuss further if I didn't manage to implement it
2. Okay so it does work, manually adding the object to PseudoScene using that function works very well. Though I just want to know where can I put the logic for whether or not a field can be parented to other object or not when dragged

**Reply by yasirkula** [reply]
2. You can add a CanChangeParentTo event to RuntimeHierarchy. Then call it inside HierarchyDragDropListener.OnDrop to cancel the drop operation if necessary.

**Reply by Rhitomaic** [reply]
Okay managed to implemented it :D
Thanks for your help!

**Reply by Rhitomaic** [reply]
I want to reopen this discussion again since I ran into a little trouble

So I want to add an object called "Camera" in a pseudo scene with a host transform without changing it's parent but still shows as a child, is that possible?

**Reply by yasirkula** [reply]
If you mean showing it as a child of the pseudo-scene, calling runtimeInspector.AddToPseudoScene is supposed to work.

**Reply by yasirkula** [reply]
Oh, but if it has rootTransform it won't work yeah. In that case, you could reparent camera to old parent immediately afterwards. PS. Drag&dropping the camera would change its parent again. I think rootTransform mustn't be set in this scenario.

**Reply by Rhitomaic** [reply]
I tried doing that and it just straight up not show in the hierarchy, will try making the base pseudo-scene to not have a host transform

**Reply by Rhitomaic** [reply]
Okay, using a normal pseudo-scene is fine, thank you
